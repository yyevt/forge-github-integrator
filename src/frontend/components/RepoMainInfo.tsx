import {Heading, Inline, Link, Lozenge, Tag, TagGroup, Text} from "@forge/react";
import React, {useMemo} from "react";
import {RepositoryPresentationDto} from "../../common/PresentationTypes";

export const RepoMainInfo = ({repo}: { repo: RepositoryPresentationDto }) => {
    const formattedCreatedDate = useMemo(
        () => formatDate(repo.createdAt),
        [repo]
    );

    return (
        <>
            <Heading as="h2">
                <Inline>
                    <Link href={repo.url}>{repo.name}</Link>
                    <Tag text={repo.visibility.toLowerCase()}/>
                    <Lozenge appearance={repo.isDisabled ? "removed" : "success"}>{repo.isDisabled ? "Disabled" : "Not disabled"}</Lozenge>
                    <Lozenge appearance={repo.isArchived ? "removed" : "success"}>{repo.isArchived ? "Archived" : "Not archived"}</Lozenge>
                </Inline>
            </Heading>
            <Text>Created: {formattedCreatedDate}</Text>
            <Text>Owner: {repo.ownerLogin}</Text>
            {repo.languages.length > 0 && (
                <TagGroup>
                    {repo.languages.map(lang => <Tag text={lang}/>)}
                </TagGroup>
            )}
        </>
    );
}

const formatDate = (dateValue: string) => new Date(dateValue).toDateString();