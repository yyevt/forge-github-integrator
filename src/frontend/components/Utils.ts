import {showFlag} from "@forge/bridge";
import {FlagType} from "@forge/bridge/out/flag/flag";

export const showToast = (id: string, title: string, type: FlagType, description: string): void => {
    showFlag({
        id: `${id}_${Math.random()}`,
        title,
        type,
        description,
        isAutoDismiss: true
    });
};