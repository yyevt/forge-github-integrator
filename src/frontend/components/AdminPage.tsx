import {
    ErrorMessage,
    Form,
    FormFooter,
    FormHeader,
    FormSection,
    HelperMessage,
    Label,
    LoadingButton,
    RequiredAsterisk,
    Textfield,
    useForm,
    ValidMessage
} from "@forge/react";
import React, {useState} from "react";
import {ForgeGateway} from "../services/ForgeGateway";
import {showToast} from "./Utils";

type FormInputs = {
    token: string;
}

export const AdminPage = () => {
    const {register, getFieldId, handleSubmit, formState: {errors, touchedFields}} = useForm<FormInputs>();
    const [isSubmitting, setSubmitting] = useState(false);

    const onSubmit = async (data: FormInputs): Promise<void> => {
        setSubmitting(true);
        try {
            await new ForgeGateway().saveToken(data.token);
            showToast("success", "Success!", "success", "Token was successfully saved.");
        } catch (e) {
            console.error(e);
            showToast("err", "Error", "error", (e instanceof Error ? e.message : `Unknown error occurred: ${e}`));
        } finally {
            setSubmitting(false);
        }
    };

    const isTokenFieldValid = !errors["token"];
    const isTokenFieldTouched = Boolean(touchedFields["token"]);

    return (
        <>
            <Form onSubmit={handleSubmit(onSubmit)}>
                <FormHeader title="Please enter GitHub token">
                    Required fields are marked with an asterisk <RequiredAsterisk/>
                </FormHeader>
                <FormSection>
                    <Label labelFor={getFieldId("token")}>GitHub token <RequiredAsterisk/></Label>
                    <Textfield placeholder="Paste token here" {...register("token", {
                        required: true,
                        pattern: /^(gh[pousr]_[a-zA-Z0-9]{36,255})$/
                    })} />
                    {isTokenFieldValid ? (
                        isTokenFieldTouched ? (
                            <ValidMessage>Valid token format</ValidMessage>
                        ) : (
                            <HelperMessage>You can use classic tokens or fine-grained ones.</HelperMessage>
                        )
                    ) : (
                        <ErrorMessage>Invalid token format</ErrorMessage>
                    )}
                </FormSection>
                <FormFooter>
                    <LoadingButton appearance="primary" type="submit" isLoading={isSubmitting}>
                        Save
                    </LoadingButton>
                </FormFooter>
            </Form>
        </>
    );
}