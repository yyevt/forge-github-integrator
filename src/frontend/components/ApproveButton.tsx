import React, {useCallback, useState} from "react";
import {LoadingButton} from "@forge/react";
import {showToast} from "./Utils";
import {ForgeGateway} from "../services/ForgeGateway";
import {PullPresentationDto} from "../../common/PresentationTypes";

export const ApproveButton = ({pullRequest}: { pullRequest: PullPresentationDto }) => {
    const [isApproving, setIsApproving] = useState(false);

    const approve = useCallback(async (): Promise<void> => {
        setIsApproving(true);
        try {
            await new ForgeGateway().approvePullRequest(pullRequest);
            showToast("success", "Success!", "success", "Pull.ts request was successfully approved.");
        } catch (e) {
            console.error(e);
            showToast("err", "Error", "error", (e instanceof Error ? e.message : `Unknown error occurred: ${e}`));
        } finally {
            setIsApproving(false);
        }
    }, [pullRequest]);

    return (
        <LoadingButton
            isLoading={isApproving}
            onClick={approve}
        >
            Approve
        </LoadingButton>
    );
};