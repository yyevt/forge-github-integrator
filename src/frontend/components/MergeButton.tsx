import React, {useCallback, useState} from "react";
import {LoadingButton} from "@forge/react";
import {showToast} from "./Utils";
import {ForgeGateway} from "../services/ForgeGateway";
import {PullPresentationDto} from "../../common/PresentationTypes";

export const MergeButton = ({pullRequest, isDisabled}: { pullRequest: PullPresentationDto, isDisabled: boolean }) => {
    const [isMerging, setIsMerging] = useState(false);

    const merge = useCallback(async (): Promise<void> => {
        setIsMerging(true);
        try {
            await new ForgeGateway().mergePullRequest(pullRequest);
            showToast("success", "Success!", "success", "Pull.ts request was successfully merged.");
        } catch (e) {
            console.error(e);
            showToast("err", "Error", "error", (e instanceof Error ? e.message : `Unknown error occurred: ${e}`));
        } finally {
            setIsMerging(false);
        }
    }, [pullRequest]);

    return (
        <LoadingButton
            isLoading={isMerging}
            isDisabled={isDisabled}
            onClick={merge}
        >
            Merge
        </LoadingButton>
    );
};