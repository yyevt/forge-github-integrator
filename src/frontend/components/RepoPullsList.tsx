import {ButtonGroup, DynamicTable, EmptyState, Link, Lozenge} from "@forge/react";
import React, {useMemo} from "react";
import {HeadType} from "@atlaskit/dynamic-table/types";
import {ApproveButton} from "./ApproveButton";
import {MergeButton} from "./MergeButton";
import {PullPresentationDto} from "../../common/PresentationTypes";

type RepoPullsListProps = {
    readonly pulls: ReadonlyArray<PullPresentationDto>;
    readonly pullsTotalCount: number;
    readonly isMergeAllowed: boolean;
    readonly onPullMerge: () => void;
}

export const head: HeadType = {
    cells: [
        {key: "issue", content: "Issue", isSortable: true},
        {key: "isIssueClosed", content: "Is Closed", isSortable: false},
        {key: "title", content: "Title", isSortable: true},
        {key: "author", content: "Author", isSortable: true},
        {key: "state", content: "State", isSortable: true},
        {key: "isDraft", content: "Is Draft", isSortable: false},
        {key: "isClean", content: "Is Clean", isSortable: false},
        {key: "actions", content: "Actions", isSortable: false},
    ],
};

export const RepoPullsList = ({pulls, isMergeAllowed, pullsTotalCount, onPullMerge}: RepoPullsListProps) => {
    const tableRows = useMemo(
        () => createTableRows(pulls, isMergeAllowed, onPullMerge),
        [pulls, isMergeAllowed, onPullMerge]
    );

    return <>
        {pulls.length === 0 ? (
            <EmptyState header="Have no opened pull requests"/>
        ) : (
            <DynamicTable
                caption={`Repository Open PRs: ${pullsTotalCount}`}
                head={head}
                rows={tableRows}
                rowsPerPage={10}
            />
        )}
    </>;
};

const createTableRows = (pulls: ReadonlyArray<PullPresentationDto>, isMergeAllowed: boolean, onPullMerge: () => void) => {
    return pulls.map((pull, index) => {

        const isPullNonMergeable = isPullClosed(pull)
            || isPullMerged(pull)
            || isPullDraft(pull)
            || isPullNotClean(pull)
            || isRepoDisabled(pull)
            || isRepoArchived(pull);

        return ({
            key: `row-${index}-${pull.url}`,
            cells: [
                {content: <Link href={`/browse/${pull.issueKey}`}>{pull.issueKey}</Link>},
                {content: <Lozenge appearance="moved">{pull.issueIsClosed ? "Yes" : "No"}</Lozenge>},
                {content: <Link href={pull.url}>{pull.title}</Link>},
                {content: pull.authorLogin},
                {content: <Lozenge appearance="inprogress">{pull.state}</Lozenge>},
                {content: <Lozenge appearance="moved">{isPullDraft(pull) ? "Yes" : "No"}</Lozenge>},
                {content: <Lozenge appearance="moved">{isPullNotClean(pull) ? "No" : "Yes"}</Lozenge>},
                {
                    content: (
                        <ButtonGroup>
                            <ApproveButton
                                pullRequest={pull}
                            />
                            <MergeButton
                                pullRequest={pull}
                                isDisabled={!isMergeAllowed || isPullNonMergeable}
                                onPullMerge={onPullMerge}/>
                        </ButtonGroup>
                    )
                }
            ]
        });
    });
}

const isPullClosed = (pull: PullPresentationDto) => pull.state !== 'OPEN';
const isPullMerged = (pull: PullPresentationDto) => pull.merged;
const isPullDraft = (pull: PullPresentationDto) => pull.isDraft;
const isPullNotClean = (pull: PullPresentationDto) => pull.mergeable != "MERGEABLE";
const isRepoDisabled = (pull: PullPresentationDto) => pull.repoIsDisabled;
const isRepoArchived = (pull: PullPresentationDto) => pull.repoIsArchived;
