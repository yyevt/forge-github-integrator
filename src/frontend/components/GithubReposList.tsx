import {Box, EmptyState, Stack} from "@forge/react";
import React from "react";
import {RepoMainInfo} from "./RepoMainInfo";
import {RepoPullsList} from "./RepoPullsList";
import {RepositoryPresentationDto} from "../../common/PresentationTypes";

export const GithubReposList = ({repos}: { repos: ReadonlyArray<RepositoryPresentationDto> }) => {
    return (
        <>
            {repos.length === 0 ? (
                <EmptyState header="Have no repositories"/>
            ) : (
                repos.map(repo => (
                    <Box padding="space.300" backgroundColor="color.background.discovery">
                        <Stack>
                            <RepoMainInfo repo={repo}/>
                            <RepoPullsList
                                pulls={repo.pullRequests}
                                pullsTotalCount={repo.pullRequestsTotalCount}
                                isMergeAllowed={repo.isMergeCommitAllowed}
                            />
                        </Stack>
                    </Box>
                ))
            )}
        </>
    );
}