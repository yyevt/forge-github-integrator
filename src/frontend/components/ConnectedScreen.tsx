import React, {useEffect, useState} from "react";
import {SectionMessage, Spinner, Stack, Text} from "@forge/react";
import {ForgeGateway} from "../services/ForgeGateway";
import {GithubReposList} from "./GithubReposList";
import {RepositoryPresentationDto} from "../../common/PresentationTypes";

export const ConnectedScreen = () => {
    const [isLoading, setIsLoading] = useState(false);
    const [repos, setRepos] = useState<ReadonlyArray<RepositoryPresentationDto>>([]);
    const [fetchReposError, setFetchReposError] = useState("");

    useEffect(() => {
        (async () => {
            setFetchReposError("");
            setIsLoading(true);
            try {
                const info = await new ForgeGateway().getGithubReposWithPulls();
                setRepos(info);
            } catch (e) {
                setFetchReposError(e instanceof Error ? e.message : `Unknown error occurred: ${e}`);
            } finally {
                setIsLoading(false);
            }
        })();
    }, []);

    if (isLoading) {
        return <Spinner/>;
    }
    return (
        <Stack space="space.200">
            <SectionMessage appearance="warning">
                <Text>Only limited amount of repositories and pull-requests are shown: max 100 repos, max 100 pulls</Text>
            </SectionMessage>
            {fetchReposError && (
                <SectionMessage appearance="error">
                    <Text>{fetchReposError}</Text>
                </SectionMessage>
            )}
            <GithubReposList repos={repos}/>
        </Stack>
    );
}