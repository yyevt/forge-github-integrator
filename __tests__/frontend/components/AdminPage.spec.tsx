import React from "react";
import renderer from "react-test-renderer";
import {AdminPage} from "../../../src/frontend/components/AdminPage";

jest.mock("@forge/bridge", () => ({
    invoke: jest.fn()
}));

describe("AdminPage", () => {

    it("renders correctly", () => {
        const reactRenderer = renderer.create(<AdminPage/>);

        expect(reactRenderer.toJSON()).toMatchSnapshot();
    });
});