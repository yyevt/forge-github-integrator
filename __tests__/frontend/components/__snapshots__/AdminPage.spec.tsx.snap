// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AdminPage renders correctly 1`] = `
<Form
  onSubmit={[Function]}
>
  <FormHeader
    title="Please enter GitHub token"
  >
    Required fields are marked with an asterisk 
    <RequiredAsterisk />
  </FormHeader>
  <FormSection>
    <Label
      labelFor="form-:r0:-token"
    >
      GitHub token 
      <RequiredAsterisk />
    </Label>
    <Textfield
      aria-invalid={false}
      id="form-:r0:-token"
      isInvalid={false}
      name="token"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Paste token here"
    />
    <HelperMessage>
      You can use classic tokens or fine-grained ones.
    </HelperMessage>
  </FormSection>
  <FormFooter>
    <LoadingButton
      appearance="primary"
      isLoading={false}
      type="submit"
    >
      Save
    </LoadingButton>
  </FormFooter>
</Form>
`;
